docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,500 @@
+to `L` under constraints `C`:
+  - If `R<B0, ..., Bj>` is the superclass of `R<B0, ..., Bj>` and `P<M0, ...,
+Mk>` is a subtype match for `Q<N0, ..., Nj>` with respect to `L` under
+constraints `C`.
+  - Or `R<B0, ..., Bj>` is one of the interfaces implemented by `P<M0, ..., Mk>` 
+(considered in lexical order) and `R<B0, ..., Bj>` is a subtype match for `Q<N0,
+..., Nj>` with respect to `L` under constraints `C`.
+  - Or `R<B0, ..., Bj>` is a mixin into `P<M0, ..., Mk>` (considered in lexical
+order) and `R<B0, ..., Bj>` is a subtype match for `Q<N0, ..., Nj>` with respect
+to `L` under constraints `C`.
+- A type `P` is a subtype match for `Function` with respect to `L` under no constraints:
+  - If `P` implements a call method.
+  - Or if `P` is a function type.
+- A type `P` is a subtype match for a type `Q` with respect to `L` under
+constraints `C`:
+  - If `P` is an interface type which implements a call method of type `F`, and
+  `F` is a subtype match for a type `Q` with respect to `L` under constraints
+  `C`.
+- A function type `(M0,..., Mn, [M{n+1}, ..., Mm]) -> R0` is a subtype match for
+  a function type `(N0,..., Nk, [N{k+1}, ..., Nr]) -> R1` with respect to `L`
+  under constraints `C0 + ... + Cr + C`
+  - If `R0` is a subtype match for a type `R1` with respect to `L` under
+  constraints `C`:
+  - If `n <= k` and `r <= m`.
+  - And for `i` in `0...r`, `Mi` is a subtype match for `Ni` with respect to `L`
+  under constraints `Ci`.
+- Function types with named parameters are treated analogously to the positional
+  parameter case above.
+- A generic function type `<T0 extends B0, ..., Tn extends Bn>F0` is a subtype
+match for a generic function type `<S0 extends B0, ..., Sn extends Bn>F1` with
+respect to `L` under constraints `Cl`:
+  - If `F0[Z0/T0, ..., Zn/Tn]` is a subtype match for `F0[Z0/S0, ..., Zn/Sn]`
+with respect to `L` under constraints `C`, where each `Zi` is a fresh type
+variable with bound `Bi`.
+  - And `Cl` is `C` with each constraint replaced with its closure with respect
+    to `[Z0, ..., Zn]`.
+
+## Expression inference
+
+Expression inference uses information about what constraints are imposed on the
+expression by the context in which the expression occurs.  An expression may
+occur in a context which provides no typing expectation, in which case there is
+no contextual information.  Otherwise, the contextual information takes the form
+of a type schema which describes the structure of the type to which the
+expression is required to conform by its context of occurrence.
+
+The primary function of expression inference is to determine the parameter and
+return types of closure literals which are not explicitly annotated, and to fill
+in elided type variables in constructor calls, generic method calls, and generic
+literals.
+
+### Expectation contexts
+
+A typing expectation context (written using the meta-variables `J` or `K`) is
+either a type schema `P`, or an empty context `_`.
```

Comment:
Does an empty context behave differently from a type schema of `?` ?

If so, what is the rationale?  If not, can we simplify the exposition by saying `?` instead of `_` everywhere below?